flocker/control/httpapi.py

```diff
@@ -185,6 +185,134 @@ def saved(ignored):
         saving.addCallback(saved)
         return saving
 
+    @app.route("/datasets/<dataset_id>", methods=['POST'])
+    @user_documentation(
+        """
+        Update an existing dataset.
+        """,
+        examples=[
+            u"update dataset with primary",
+        ]
+    )
+    @structured(
+        # XXX Validation of the supplied primary address is performed by input
+        # schema validation. Any further checks required?
+        inputSchema={'$ref': '/v1/endpoints.json#/definitions/datasets'},
+        outputSchema={'$ref': '/v1/endpoints.json#/definitions/datasets'},
+        schema_store=SCHEMAS
+    )
+    def update_dataset(self, dataset_id, primary=None, maximum_size=None,
+                       metadata=None):
+        """
+        Update an existing dataset in the cluster configuration.
+
+        :param unicode dataset_id: The unique identifier of the dataset.  This
+            is a string giving a UUID (per RFC 4122).
+
+        :param unicode primary: The address of the node to which the dataset
+            will be moved.
+
+        :param int maximum_size: The maximum number of bytes the dataset will
+            be capable of storing.  This may be optional or required depending
+            on the dataset backend.
+
+        :param dict metadata: A small collection of unicode key/value pairs to
+            associate with the dataset.  These items are not interpreted.  They
+            are only stored and made available for later retrieval.  Use this
+            for things like human-friendly dataset naming, ownership
+            information, etc.
+
+        :return: A ``dict`` describing the dataset which has been added to the
+            cluster configuration or giving error information if this is not
+            possible.
+        """
+        # XXX Changing volume size and metadata via the API is beyond the scope
+        # of this issue. Decide how to react.
+        # if maximum_size is None:
+        #     raise NotImplemented()
+
+        # if metadata is None:
+        #     raise NotImplemented()
+
+        # Get the current configuration
+        # deployment = self.persistence_service.get()
+
+        # Find all manifestations of dataset
+        # manifestations = manifestations_from_deployment(deployment, dataset_id)
+
+        # if not manifestations:
+        #     raise NotFound(
+        #         'Dataset not found. '
+        #         'There are no manifestations of the requested dataset. '
+        #         'dataset_id: {}'.format(dataset_id)
+        #     )
+
+        # Lookup the node that has a primary Manifestation (if any)
+
+        # expected_manifestation = Manifestation(dataset=dataset, primary=True)
+        # primary_nodes = list(
+        #     node for node in deployment.nodes if node.hostname == primary
+        # )
+        # if len(primary_nodes) == 0:
+        #     # No primary node
+        #     # raise something here?
```

Comment:
No primary node is a valid configuration, even if we possibly don't support it yet. E.g. in OpenStack or EBS backend you might have dataset that exists but isn't attached to any VM.